### PR TITLE
fix(agw): Modified code to port teid has list to protobuf in sgw_s8 task

### PR DESCRIPTION
**Pull request by rsarwad**
## Summary
fix(agw): Modified code to port teid has list to protobuf in sgw_s8 task
## Test Plan
Executed only unit test suite. For in-bound roaming, End2End scenarios used to be tested in Tera VM setup. Since we are not using the Tera VM setup. We couldn't validate End2End scenario.

## Additional Information
Once federated integ_test supports mocked external PGW, we can validate the End2End scenarios for inbound roaming UE

### CODE REVIEW

**Comment by ajahl** [comment]
Can you fix the linter warnings and rebase your PR?

**Comment by ajahl** [comment]
@rsarwad Can you fix the linter issues, please?

File: lte/gateway/c/core/oai/include/proto_map.hpp [hunk 1]
```diff
@@ -170,15 +170,15 @@ struct proto_map_s {
   **              the map. If key does not exists returns error             **
   **                                                                        **
   ***************************************************************************/
-  proto_map_rc_t remove(const keyT key) {
+  proto_map_rc_t remove(const keyT key, bool free_an_entry = true) {
```
Comment (pruthvihebbani) [inline lte/gateway/c/core/oai/include/proto_map.hpp]: What is the need of this flag? Shouldn`t the entry be freed always?

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp [hunk 2]
```diff
@@ -120,11 +120,10 @@ void sgw_remove_sgw_bearer_context_information(sgw_state_t* sgw_state,
 sgw_eps_bearer_context_information_t* sgw_get_sgw_eps_bearer_context(
     teid_t teid) {
   OAILOG_FUNC_IN(LOG_SGW_S8);
-  sgw_eps_bearer_context_information_t* sgw_bearer_context_info = NULL;
-  hash_table_ts_t* state_imsi_ht = get_sgw_ue_state();
+  sgw_eps_bearer_context_information_t* sgw_bearer_context_info = nullptr;
+  map_uint32_sgw_eps_bearer_context_t* state_teid_ht = get_s8_state_teid_map();
```
Comment (pruthvihebbani) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp]: NULL check for state_teid_ht

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp [hunk 3]
```diff
@@ -110,16 +117,27 @@ void SgwStateManager::free_state() {
...
+  if (state_cache_p->temporary_create_session_procedure_id_map.map) {
+    if (state_cache_p->temporary_create_session_procedure_id_map
+            .destroy_map() != PROTO_MAP_OK) {
+      OAILOG_ERROR(LOG_SGW_S8,
+                   "An error occurred while destroying "
+                   "temporary_create_session_procedure_id_map ");
+    }
+  }
+  state_cache_p->temporary_create_session_procedure_id_map.map = nullptr;
+  free_cpp_wrapper((void**)&state_cache_p);
```
[the remark this answers is not in the document]
Comment (pruthvihebbani) [reply lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp]: Fix the warning

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp [hunk 4]
```diff
@@ -130,5 +148,14 @@ status_code_e SgwStateManager::read_ue_state_from_db() {
 sgw_state_t* SgwStateManager::get_state(bool read_from_db) {
   return state_cache_p;
 }
+
+map_uint32_sgw_eps_bearer_context_t* SgwStateManager::get_s8_state_teid_map() {
+  AssertFatal(
```
Comment (pruthvihebbani) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp:137]: Use Error log instead of assertFatal

**Review by pruthvihebbani — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/include/proto_map.hpp (hunk 1 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/include/proto_map.hpp]: If flag is set, an entry for corresponding key would be removed and also contents are freed.
But in this PR, sgw_context would be initially in map, temporary_create_session_procedure_id_map. On reception of create session response, sgw_context would be removed from temporary_create_session_procedure_id_map and inserted into state_tied_map

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp (hunk 4 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp:137]: It's mandatory to have this flag, is_initialized to be initialized. So I think assertFatal should be fine

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.hpp [hunk 5]
```diff
@@ -21,10 +21,11 @@ limitations under the License.
 
 namespace {
 constexpr int SGW_STATE_CONTEXT_HT_MAX_SIZE = 512;
-constexpr char S11_BEARER_CONTEXT_INFO_HT_NAME[] =
+constexpr char S11_BEARER_CONTEXT_INFO_MAP_NAME[] =
     "s11_bearer_context_information_htbl";
 constexpr char SGW_STATE_TABLE_NAME[] = "sgw_state";
 constexpr char SGW_TASK_NAME[] = "SGW";
+constexpr char SGW_S8_CSR_PROC_ID_MAP[] = "sgw_s8_csr_proc_id_map";
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.hpp]:
```suggestion
constexpr char SGW_S8_CSR_PROC_ID_MAP_NAME[] = "sgw_s8_csr_proc_id_map";
```

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp (hunk 4 above)
Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp:137]: Yeah, if the state is not cleanly initialized from Redis, the service should fail to start up.

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp [hunk 6]
```diff
@@ -1833,11 +1839,9 @@ sgw_eps_bearer_context_information_t* update_sgw_context_to_s11_teid_map(
   /* Once sgw_s8_teid is obtained from orc8r, move sgw_eps_bearer_context
    * from temporary_create_session_procedure_id hashlist  to sgw_teid hashlist
    */
-  sgw_eps_bearer_context_information_t* sgw_context_p = NULL;
-  hashtable_ts_remove(
-      sgw_state->temporary_create_session_procedure_id_htbl,
-      (const hash_key_t)session_rsp_p->temporary_create_session_procedure_id,
-      (void**)&sgw_context_p);
+  sgw_eps_bearer_context_information_t* sgw_context_p = nullptr;
+  sgw_state->temporary_create_session_procedure_id_map.get(
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp:1822]: What's the purpose of getting this entry before removing it?

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp [hunk 7]
```diff
@@ -1846,17 +1850,30 @@ sgw_eps_bearer_context_information_t* update_sgw_context_to_s11_teid_map(
     OAILOG_FUNC_RETURN(LOG_SGW_S8, sgw_context_p);
   }
 
+  if (sgw_state->temporary_create_session_procedure_id_map.remove(
+          session_rsp_p->temporary_create_session_procedure_id, false) !=
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp]: Why do we need the remove with `false` flag here? In the original logic the entry is being removed and freed from the hash table.

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state.cpp [hunk 8]
```diff
@@ -59,12 +59,16 @@ void put_sgw_ue_state(sgw_state_t* sgw_state, imsi64_t imsi64) { return; }
...
+    return;
+  }
+  sgw_eps_bearer_context_information_t* sgw_eps_context =
+      reinterpret_cast<sgw_eps_bearer_context_information_t*>(*ptr);
+  if (sgw_eps_context) {
+    sgw_free_pdn_connection(&sgw_eps_context->pdn_connection);
+    delete_pending_procedures(sgw_eps_context);
   }
-  free_wrapper((void**)sgw_eps_context);
+  free_wrapper(reinterpret_cast<void**>(ptr));
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state.cpp]:
```suggestion
  free_cpp_wrapper(reinterpret_cast<void**>(ptr));
```

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp (hunk 6 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp:1822]: While handling create session request, sgw_context is created and inserted into map,  temporary_create_session_procedure_id_map. Because for roaming users, sgw_s11_teid would be allocated by external PGW, shall be received in create session response message. So on reception of create session response, sgw_context is removed from map, but not freed. 
So first we get the sgw_context, remove from map, temporary_create_session_procedure_id_map and insert into state_teid_map

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp (hunk 7 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp]: In the original code, only entry is removed but not freed from hash table. This is taken care by API, hashtable_ts_remove() which only removes an entry from hash table but doesn't free. For API, hashtable_ts_free() free_wrapper is called as here https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/lib/hashtable/hashtable.c#L664
But same is not true for hashtable_ts_remove(), https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/lib/hashtable/hashtable.c#L691
The original code is here, https://github.com/magma/magma/pull/14006/files#diff-89dd268674e287dc1dc45f2515f69e41f864be6074f3e03529324cb1eeecb374L1837

To implement same behavior, I added new flag.

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state.cpp (hunk 8 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state.cpp]: In this PR, memory for sgw_eps_bearer_context_information_t is allocated with calloc. But in other PR, memory is allocated with new(). It's here, https://github.com/magma/magma/pull/13895/files#diff-89dd268674e287dc1dc45f2515f69e41f864be6074f3e03529324cb1eeecb374R189
It's shall be taken care while rebasing.

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp [hunk 9]
```diff
@@ -110,16 +117,27 @@ void SgwStateManager::free_state() {
...
-        "An error occurred while destroying SGW s11_bearer_context_information "
-        "hashtable");
+  if (s8_state_teid_map.map) {
+    if (s8_state_teid_map.destroy_map() != PROTO_MAP_OK) {
+      OAILOG_ERROR(LOG_SGW_S8,
+                   "An error occurred while destroying "
+                   "temporary_create_session_procedure_id_map ");
+    }
   }
+  s8_state_teid_map.map = nullptr;
```
Comment (ajahl) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp]: Is there a reason that you can not set `map` to `nullptr` in function `destroy_map()`?

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp [hunk 10]
```diff
@@ -110,16 +117,27 @@ void SgwStateManager::free_state() {
...
-  free_wrapper((void**)&state_cache_p);
+  if (state_cache_p->temporary_create_session_procedure_id_map.map) {
+    if (state_cache_p->temporary_create_session_procedure_id_map
+            .destroy_map() != PROTO_MAP_OK) {
+      OAILOG_ERROR(LOG_SGW_S8,
+                   "An error occurred while destroying "
+                   "temporary_create_session_procedure_id_map ");
+    }
+  }
+  state_cache_p->temporary_create_session_procedure_id_map.map = nullptr;
```
Comment (ajahl) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp]: See above

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp (hunk 9 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp]: I had missed it. Thanks for bringing into notice

**Review by pruthvihebbani — APPROVED** [review]
LGTM

**Comment by pruthvihebbani** [comment]
@rsarwad Please rebase the PR

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp [hunk 11]
```diff
@@ -124,7 +124,6 @@ void SgwStateManager::free_state() {
                    "temporary_create_session_procedure_id_map ");
     }
   }
-  s8_state_teid_map.map = nullptr;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp]: Why is this removed?

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp [hunk 12]
```diff
@@ -136,7 +135,6 @@ void SgwStateManager::free_state() {
                    "temporary_create_session_procedure_id_map ");
     }
   }
-  state_cache_p->temporary_create_session_procedure_id_map.map = nullptr;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp:129]: Why is this removed?

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp (hunk 11 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp]: This is removed because map is set to nullptr within destroy_map() itself. This change is done while addressing the comment, https://github.com/magma/magma/pull/14006#discussion_r1007844896

File: lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp (hunk 12 above)
Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_state_manager.cpp:129]: This is removed because map is set to nullptr within destroy_map() itself. This change is done while addressing the comment, https://github.com/magma/magma/pull/14006#discussion_r1007844896

**Review by ssanadhya — APPROVED** [review]
Lgtm

**Comment by ssanadhya** [comment]
@rsarwad , please rebase to retrigger CI.
